### PR TITLE
[WIP] Check node sizes

### PR DESCRIPTION
**Pull request by sandreenko**

### CODE REVIEW

File: src/jit/gentree.cpp [hunk 1]
```diff
@@ -305,30 +292,30 @@ void GenTree::InitNodeSize()
     static_assert_no_msg(sizeof(GenTreeLclVar)       <= TREE_NODE_SZ_SMALL);
     static_assert_no_msg(sizeof(GenTreeLclFld)       <= TREE_NODE_SZ_SMALL);
     static_assert_no_msg(sizeof(GenTreeCC)           <= TREE_NODE_SZ_SMALL);
-    static_assert_no_msg(sizeof(GenTreeCast)         <= TREE_NODE_SZ_LARGE); // *** large node
-    static_assert_no_msg(sizeof(GenTreeBox)          <= TREE_NODE_SZ_LARGE); // *** large node
+    static_assert_no_msg(sizeof(GenTreeCast)         <= TREE_NODE_SZ_SMALL);
```
Comment (mikedn) [inline src/jit/gentree.cpp]: How do you know when it is possible to make it small? AFAIR some casts are turned into helper calls.

**Comment by sandreenko** [comment]
> How do you know when it is possible to make it small?

I do not know (for all operands), so this pr is checking this. I have checked a few converted operands manually but not all.

> AFAIR some casts are turned into helper calls.

I am aware of 3 methods that turn nodes: `GenTree::ReplaceWith`, `GenTree::SetOper`, and `GenTree::SetOperRaw`. The first two check sizes and we will see asserts if something went wrong. The third does not check anything (why?).

And we have 3 types of nodes: 
1) small nodes;
2) large nodes;
3) large nodes that was created from with small node GenTree* types with opcode from a large node, like:
`new (this, LargeOpOpcode()) GenTreeLclVar(type, lnum DEBUGARG(ILoffs) DEBUGARG(/*largeNode*/ true));`.

As I understand the best case for throughput/memory consumption would be if we create all nodes which `sizeof(GenTree*)` is small as 3) in necessary cases.

**Comment by mikedn** [comment]
> I do not know (for all operands), so this pr is checking this. I have checked a few converted operands manually but not all.

I think you'll need to manually check everything. I don't think we can rely just on test coverage to check something like this, test coverage might not be enough and getting this wrong will have dire consequences in release builds.

> I am aware of 3 methods that turn nodes: GenTree::ReplaceWith, GenTree::SetOper, and GenTree::SetOperRaw. The first two check sizes and we will see asserts if something went wrong. The third does not check anything (why?).

I don't think `SetOperRaw` was intended to be used directly. But the whole thing is pretty confusing, e.g. when you're supposed to use `SetOper` and when `ChangeOper`. IMO it would have been better if `SetOper` only changed the oper, without changing the class (e.g. `GT_SUB` -> `GT_ADD`) and `ChangeOper` changed the class as well (and all this was enforced by asserts to make sure people don't use them interchangeably like it seems to happen now). That would also made it more easier to find problematic changes in the code.

> As I understand the best case for throughput/memory consumption would be if we create all nodes which sizeof(GenTree*) is small as 3) in necessary cases.

Yes, of course. Except that 3 seems less safe. It may be better to focus on those opers that are now always large for completely bogus reasons (e.g. I have no idea why `GT_ARR_BOUNDS_CHECK` is large now, I'm 99.99% sure it doesn't have to) and leave the rest alone for now. Those that need to be large because they may be turned into calls may benefit from additional consideration:
* does that happen often enough that simply allocating a new node would be a problem
* do call nodes really have to be large? (yes, I mean it, the structure of call nodes is just horrible).

File: src/jit/gentree.cpp [hunk 2]
```diff
@@ -255,16 +250,9 @@ void GenTree::InitNodeSize()
...
-    GenTree::s_gtNodeSizes[GT_ARR_OFFSET]       = TREE_NODE_SZ_LARGE;
-    GenTree::s_gtNodeSizes[GT_RET_EXPR]         = TREE_NODE_SZ_LARGE;
+    GenTree::s_gtNodeSizes[GT_RET_EXPR]         = TREE_NODE_SZ_LARGE; // Can be bashed to GT_CALL.
     GenTree::s_gtNodeSizes[GT_OBJ]              = TREE_NODE_SZ_LARGE;
     GenTree::s_gtNodeSizes[GT_FIELD]            = TREE_NODE_SZ_LARGE;
-    GenTree::s_gtNodeSizes[GT_STMT]             = TREE_NODE_SZ_LARGE;
-    GenTree::s_gtNodeSizes[GT_CMPXCHG]          = TREE_NODE_SZ_LARGE;
-    GenTree::s_gtNodeSizes[GT_QMARK]            = TREE_NODE_SZ_LARGE;
-    GenTree::s_gtNodeSizes[GT_LEA]              = TREE_NODE_SZ_LARGE;
-    GenTree::s_gtNodeSizes[GT_STORE_OBJ]        = TREE_NODE_SZ_LARGE;
```
Comment (mikedn) [inline src/jit/gentree.cpp]: `GT_STORE_OBJ` has to be large, like `GT_OBJ`. I happen to have a PR that makes it small.

**Comment by sandreenko** [comment]
> I think you'll need to manually check everything. I don't think we can rely just on test coverage to check something like this, test coverage might not be enough and getting this wrong will have dire consequences in release builds.

My plan of attacking this issue is:
1. a pr with types/comments/names fixes (subset of this PR);
2. measure TP and PIN impact of this PR to get the upper boundary for the possible impact;
if the second prove that we want to invest in it then:
3 a. start manually changing sizes one by one;
3 b. or refactor `ReplaceWith` to use only pointers (the biggest issue here is with argInfo as always), clean `SetOper/SetOperRaw/ChangeOper` to make 3 a. obvious.


I do not expect 3 a.b to be in 3.0 release, so it is future planning.

**Comment by sandreenko** [comment]
So I measured x64 Release System.Private.CoreLib and expected to save 43.883.136 bytes with my change in theory (based on the number of each tree type in the constructor) and the measurements showed 39.995.776.
But it gave ~1.8% allocation size decrease (2.185.798.670->2.145.802.894 bytes, ASTNode |  412744848 |  18.88% -> ASTNode |  372749072 |  17.37%) and nothing in PIN results.

**Comment by AndyAyersMS** [comment]
Not surprising to see instruction count stay the same, presumably those extra bytes were never accessed.

TP savings from shrinking allocations will more likely be indirect (better cache density, fewer private pages, etc) and will be in cycles, not instructions.

**Pull request closed without merging**